### PR TITLE
Changes Related to VM Name as optional Parameter and Filtering ListVMs to Filter Using the Datacenter, Cluster and Folder

### PR DESCRIPTION
**Pull request by kunal-pmj**

### CODE REVIEW

File: src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java [hunk 1]
```diff
@@ -35,27 +35,32 @@
...
+		
+	    Cluster clusterService = stubFactory.createStub(Cluster.class,
+	        sessionStubConfig);
+	    Set<String> clusters = Collections.singleton(clusterName);
+	    ClusterTypes.FilterSpec.Builder clusterFilterBuilder = 
+	    		new ClusterTypes.FilterSpec.Builder().setNames(clusters);
+	    if( null != datacenterName){
+	    	// Get the datacenter
+	    	Set<String> datacenters = Collections.singleton(DatacenterHelper
+	        .getDatacenter(stubFactory, sessionStubConfig, datacenterName));
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java]: Formatting is a little off here. Please fix it.

File: src/main/java/vmware/samples/vcenter/helpers/FolderHelper.java [hunk 2]
```diff
@@ -38,24 +38,32 @@
...
-                    .setDatacenters(datacenters)
-                    .build();
+        FolderTypes.FilterSpec.Builder vmFolderFilterSpecBuilder =
+                new FolderTypes.FilterSpec.Builder().setNames(vmFolders);
+                    
+        if(null != datacenterName)
+        {	
+        	// Get the datacenter
+        	Set<String> datacenters = Collections.singleton(DatacenterHelper
+            .getDatacenter(stubFactory, sessionStubConfig, datacenterName));
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/helpers/FolderHelper.java]: Again, please fix formatting.

File: src/main/java/vmware/samples/vcenter/helpers/FolderHelper.java [hunk 3]
```diff
@@ -38,24 +38,32 @@
...
-            vmFolderFilterSpec);
+        		vmFolderFilterSpecBuilder.build());
         assert folderSummaries.size() > 0 : "Folder " + folderName
-                                            + "not found in datacenter: "
+                                            + " not found in datacenter: "
                                             + datacenterName;
         return folderSummaries.get(0).getFolder();
     }
+
+	public static String getFolder(StubFactory stubFactory, StubConfiguration sessionStubConfig, String vmFolderName) {
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/helpers/FolderHelper.java]: We have followed a 80 char line length limit in all the samples. Lets keep it consistent to any new changes as well.

File: src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java [hunk 4]
```diff
@@ -39,15 +46,56 @@
      * @param args command line arguments passed to the sample
      */
     protected void parseArgs(String[] args) {
-        super.parseArgs(Collections.<Option>emptyList(), args);
+        Option datacenterOption = Option.builder()
+                .longOpt("datacenter")
+                .desc("OPTIONAL: Specify the name of the Datacenter to list the Vms available in it.")
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Vms -> VMs, fix line length to 80

File: src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java [hunk 5]
```diff
@@ -39,15 +46,56 @@
...
+                .argName("DATACENTER")
+                .required(false)
+                .hasArg()
+                .build();
+            Option vmFolderOption = Option.builder()
+                .longOpt("vmfolder")
+                .desc("OPTIONAL: Specify the name of the VM Folder to list the Vms contained in it.")
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Vms -> VMs, fix line length to 80

File: src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java [hunk 6]
```diff
@@ -35,27 +35,32 @@
...
-                    .build();
-        List<ClusterTypes.Summary> clusterSummaries = clusterService.list(
-            clusterFilterSpec);
-        assert clusterSummaries.size() > 0 : "Cluster " + clusterName
-                                             + "not found in datacenter: "
-                                             + datacenterName;
-        return clusterSummaries.get(0).getCluster();
-    }
+	public static String getCluster(StubFactory stubFactory, StubConfiguration sessionStubConfig, String clusterName) {
+		// TODO Auto-generated method stub
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java]: Remove TODO

File: src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java [hunk 7]
```diff
@@ -35,27 +35,32 @@
...
-                    .setDatacenters(datacenters)
-                    .build();
-        List<ClusterTypes.Summary> clusterSummaries = clusterService.list(
-            clusterFilterSpec);
-        assert clusterSummaries.size() > 0 : "Cluster " + clusterName
-                                             + "not found in datacenter: "
-                                             + datacenterName;
-        return clusterSummaries.get(0).getCluster();
-    }
+	public static String getCluster(StubFactory stubFactory, StubConfiguration sessionStubConfig, String clusterName) {
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/helpers/ClusterHelper.java]: Fix line length to 80

File: src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java [hunk 8]
```diff
@@ -39,15 +46,56 @@
...
 
     protected void setup() throws Exception {
         this.vmService = vapiAuthHelper.getStubFactory().createStub(VM.class, sessionStubConfig);
     }
 
     protected void run() throws Exception {
-        List<Summary> vmList = this.vmService.list(new VMTypes.FilterSpec.Builder().build());
+    	com.vmware.vcenter.VMTypes.FilterSpec.Builder bldr = new VMTypes.FilterSpec.Builder();
+    	if(null != this.datacenterName && !this.datacenterName.isEmpty()){
+    		bldr.setDatacenters(Collections.singleton(DatacenterHelper.getDatacenter(this.vapiAuthHelper.getStubFactory(), this.sessionStubConfig, this.datacenterName)));
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Remove tab characters. Usually, its 4 spaces after the "if"

Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Fix line length

File: src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java [hunk 9]
```diff
@@ -39,15 +46,56 @@
...
+    	}
+    	if(null != this.clusterName && !this.clusterName.isEmpty()) {
+    		bldr.setClusters(Collections.singleton(ClusterHelper.getCluster(this.vapiAuthHelper.getStubFactory(), sessionStubConfig, this.clusterName)));
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Fix line length

File: src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java [hunk 10]
```diff
@@ -39,15 +46,56 @@
...
+    	}    	
+    	if(null != this.vmFolderName && !this.vmFolderName.isEmpty())
+    	{
+    		bldr.setFolders(Collections.singleton(FolderHelper.getFolder(this.vapiAuthHelper.getStubFactory(), sessionStubConfig, this.vmFolderName)));
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/list/ListVMs.java]: Fix line length

File: src/main/java/vmware/samples/vcenter/vm/create/basicvm/CreateBasicVM.java [hunk 11]
```diff
@@ -63,28 +64,35 @@
     protected void parseArgs(String[] args) {
         Option datacenterOption = Option.builder()
             .longOpt("datacenter")
-            .desc("The name of the datacenter on which to create the vm.")
+            .desc("The name of the datacenter in which to create the vm.")
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/create/basicvm/CreateBasicVM.java]: Why "in" ?

File: src/main/java/vmware/samples/vcenter/vm/create/exhaustivevm/CreateExhaustiveVM.java [hunk 12]
```diff
@@ -306,7 +316,7 @@ private void createExhaustiveVM(VMTypes.PlacementSpec vmPlacementSpec,
                 .build(),
             new DeviceTypes.EntryCreateSpec.Builder(DeviceTypes.Type.ETHERNET)
                 .build());
-
+        String vmName = (null == this.vmName || this.vmName.isEmpty())? EXHAUSTIVE_VM_NAME :this.vmName;
```
Comment (vinpai) [inline src/main/java/vmware/samples/vcenter/vm/create/exhaustivevm/CreateExhaustiveVM.java]: Fix line length

File: src/main/java/vmware/samples/vcenter/vm/create/basicvm/CreateBasicVM.java (hunk 11 above)
Comment (kunal-pmj) [inline src/main/java/vmware/samples/vcenter/vm/create/basicvm/CreateBasicVM.java]: Datacenteres would hold the Clusters, servers etc. A VM would would be hosted in it.

**Pull request closed without merging**